castle/cms/browser/controlpanel/archival.py

```diff
@@ -173,56 +173,48 @@ def save(self):
-            page = 1
-        if page <= 0:
-            page = 1
-        per_page = 1000
-        page_start = (page - 1) * per_page
-        page_end = page * per_page
-        s3_per_page = 1000
-
-        # this is so we don't kill the client with a large amount of objects being sent that way
-        # note: s3 api doesn't do paging like the client would expect to render. the pages defined
-        # for the api are basically equivalent to 1000 item max chunks retreived from the bucket.
-        #
-        # an improvement on this would probably be caching the list with every publish to the
-        # archive and having somewhere to manually update the cache. TODO.
-
-        object_itr = self.bucket.objects.page_size(count=s3_per_page).filter(Prefix=base_path)
-        i = 0
-        for summary in object_itr:
-            i += 1
-            if i < page_start:
-                continue
-            key = summary.key
-            path = key[len(archival.CONTENT_KEY_PREFIX):]
-            result.append({
-                'path': path,
-                'id': path.rstrip('/').split('/')[-1],
-                'is_folder': False,
-                'url': '{endpoint_url}/{bucket}/{key}'.format(
-                    endpoint_url=self.s3.meta.client.meta.endpoint_url,
-                    bucket=self.bucket.name,
-                    key=key)
-            })
-            if i >= page_end:
-                break
-        return result
+        folderresults = []
+        fileresults = []
+        paginator = self.bucket.meta.client.get_paginator('list_objects')
+        for resp in paginator.paginate(Bucket=self.bucket.name, Delimiter='/', Prefix=base_path):
+            for prefix in resp.get('CommonPrefixes', []):
+                key = prefix['Prefix'][len(archival.CONTENT_KEY_PREFIX):]
+                folderresults.append({
+                    'path': key,
+                    'id': key.rstrip('/').split('/')[-1],
+                    'is_folder': True,
+                    'url': '{endpoint_url}/{bucket}/{key}'.format(
+                        endpoint_url=self.s3.meta.client.meta.endpoint_url,
+                        bucket=self.bucket.name,
+                        key=key)
+                })
+            for item in resp.get('Contents', []):
+                key = item.get('Key', None)
+                if key is None:
+                    continue
+                path = key[len(archival.CONTENT_KEY_PREFIX):]
+                fileresults.append({
+                    'path': path,
+                    'id': path.rstrip('/').split('/')[-1],
+                    'is_folder': False,
+                    'url': '{endpoint_url}/{bucket}/{key}'.format(
+                        endpoint_url=self.s3.meta.client.meta.endpoint_url,
+                        bucket=self.bucket.name,
+                        key=key)
+                })
+        sorted(folderresults, key=lambda x: x['path'])
+        sorted(fileresults, key=lambda x: x['path'])
```

Comment:
sorted() doesn't modify the iterable - probably mean `folderresults = sorted(folderresults, key=lambda x: x['path'])` and `fileresults = sorted(fileresults, key=lambda x: x['path'])`